pinot-minion-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/CustomerBasedRetentionTaskGenerator.java

```diff
@@ -0,0 +1,162 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.helix.task.TaskState;
+import org.apache.pinot.common.data.Segment;
+import org.apache.pinot.common.metadata.segment.OfflineSegmentZKMetadata;
+import org.apache.pinot.controller.helix.core.minion.ClusterInfoAccessor;
+import org.apache.pinot.controller.helix.core.minion.generator.PinotTaskGenerator;
+import org.apache.pinot.controller.helix.core.minion.generator.TaskGeneratorUtils;
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.minion.PinotTaskConfig;
+import org.apache.pinot.spi.annotations.minion.TaskGenerator;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableTaskConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@TaskGenerator
+public class CustomerBasedRetentionTaskGenerator implements PinotTaskGenerator{
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CustomerBasedRetentionTaskGenerator.class);
+  private static final String TASK_TYPE = "customerBasedRetentionTask";
+  public static final String BUCKET_TIME_PERIOD_KEY = "bucketTimePeriod";
+  private static final String DEFAULT_BUCKET_PERIOD = "1d";
+  private static final String CUSTOMER_RETENTION_CONFIG = "customerRetentionConfig";
+  public static final String WINDOW_START_MS_KEY = "windowStartMs";
+  public static final String WINDOW_END_MS_KEY = "windowEndMs";
+
+  private ClusterInfoAccessor _clusterInfoAccessor;
+
+  @Override
+  public void init(ClusterInfoAccessor clusterInfoAccessor) {
+    _clusterInfoAccessor = clusterInfoAccessor;
+  }
+
+  @Override
+  public String getTaskType() {
+    return TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    // Get the segments that are being converted so that we don't submit them again
+    Set<Segment> runningSegments =
+        TaskGeneratorUtils.getRunningSegments(MinionConstants.ConvertToRawIndexTask.TASK_TYPE, _clusterInfoAccessor);
+
+    for (TableConfig tableConfig : tableConfigs) {
+      String offlineTableName = tableConfig.getTableName();
+
+      if (tableConfig.getTableType() != TableType.OFFLINE) {
+        LOGGER.warn("Skip generating task: {} for non-OFFLINE table: {}", TASK_TYPE, offlineTableName);
+        continue;
+      }
+
+      LOGGER.info("Start generating task configs for table: {} for task: {}", offlineTableName, TASK_TYPE);
+
+      // Only schedule 1 task of this type, per table
+      Map<String, TaskState> incompleteTasks =
+          TaskGeneratorUtils.getIncompleteTasks(TASK_TYPE, offlineTableName, _clusterInfoAccessor);
+      if (!incompleteTasks.isEmpty()) {
```

Comment:
when can this happen ?

Reply:
@laxmanchekka do we need this now since we are generating multiple tasks per table ?

Reply:
Yes. We still need this as we don't want to allow more than one job instance.